### DISCUSSION TITLE
How exactly is loss calculated in DeepXDE?

### DISCUSSION THREAD

**Discussion by silash35** [lululxvi/deepxde#1805] [Q&A]
> Previously, the title of the discussion was “How to replicate a model created using DeepXDE using only PyTorch?”. However, I believe that the question was too confusing and, at the same time, too broad. So I reformulated the question focusing only on how to implement the loss function.

Hello,

I am trying to build a model using PyTorch to solve a system of ordinary differential equations (ODEs) with initial conditions. I used DeepXDE with PyTorch as the backend, and it solved the problem perfectly.

However, when I try to replicate the model in PyTorch, using all the identical parameters, the model does not work correctly. I'm using the same activation function, same optimizer, same number of epochs. I tried to replicate as many things as possible and it still didn't work.

My question is: How exactly is loss calculated in DeepXDE? I mean, what mathematical formula is used?

Say, for example, the [ODE system](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/ode.system.html) from the documentation, how is the loss calculation done for each of the conditions?

**Comment by lululxvi** [comment]
> I used DeepXDE with PyTorch as the backend, and it solved the problem perfectly.

Good to know.

> My question is: How exactly is loss calculated in DeepXDE? I mean, what mathematical formula is used?

If you don't use additional tricks, the default mathematical formula in DeepXDE is what we have in the DeepXDE paper.

Why do you have to re-implement the code using PyTorch?

**Comment by silash35** [comment]
Basically, I had to re-implement the code using only PyTorch because my supervisor told me to. Probably because I'm still in the early stages of my research and learning about PINNs, so it would make sense for me to implement one without using a library like DeepXDE.

But I figured out what I was doing wrong. It was a stupid mistake when destructuring the tensors in PyTorch.

I asked this question believing that DeepXDE would be doing some other optimization not written in the paper. Some technique like [Penalizing Physical Loss Gradient](https://arxiv.org/pdf/2112.05620). But I was wrong.

Indeed, by following the instructions in the paper correctly, I was able to replicate the model created by DeepXDE with similar results.

Thank you
